dash-spv-coinjoin/src/coinjoin.rs

```diff
@@ -0,0 +1,346 @@
+use std::collections::HashMap;
+use std::ptr::null;
+
+use dash_spv_masternode_processor::chain::common::ChainType;
+use dash_spv_masternode_processor::chain::params::DUFFS;
+use dash_spv_masternode_processor::common::Block;
+use dash_spv_masternode_processor::crypto::byte_util::UInt256;
+use dash_spv_masternode_processor::tx::transaction::Transaction;
+use dash_spv_masternode_processor::util::script::ScriptType;
+use ferment_interfaces::boxed;
+
+use crate::callbacks::{GetInputValueByPrevoutHash, HasChainLock};
+use crate::messages::pool_message::PoolMessage;
+use crate::messages::pool_status::PoolStatus;
+use crate::messages::coinjoin_broadcast_tx::CoinJoinBroadcastTx;
+use crate::constants::COINJOIN_ENTRY_MAX_SIZE;
+use crate::models::InputValue;
+use crate::utils::value_from_amount::value_from_amount;
+
+#[repr(C)]
+#[derive(Debug)]
+// #[ferment_macro::export]
+pub struct CoinJoin {
+    pub opaque_context: *const std::ffi::c_void,
+    // pub chain_type: ChainType,
+    pub get_input_value_by_prevout_hash: GetInputValueByPrevoutHash,
+    pub has_chain_lock: HasChainLock,
+    map_dstx: HashMap<UInt256, CoinJoinBroadcastTx> // TODO: thread safety?
```

Comment:
Not sure what thread safety will look like yet. Don't see `mutex` being used in dash_shared_core

Reply:
At this point there no thread management in dash-shared-core. It's controlled by DashSync